design/secure-communication-design.md

```diff
@@ -0,0 +1,88 @@
+
+## Background
+
+Currently, Contour and Envoy containers are co-located within a pod.
+Current security model is relying on the fact that Contour exposes the management
+server on loopback interface which prevents any unauthorized access from outside
+the pod.
+
+With the need to move away from the [current deployment pattern of co-locating
+Contour and Envoy][0], it is necessary to secure the communication channel between
+the containers.
+
+## High level design
+
+- Use mutual TLS to secure communication channel between Contour and Envoy
+- Bootstrap phase of Contour will generate necessary TLS certificates
+
+## Detailed design
+
+### Default options for secure communication
+
+By default, mutual TLS is enabled to secure the communication between Contour and Envoy
+containers. The default `contour bootstrap` will generate the bootstrap configuration that
+enables secure communication between the containers. If `contour bootstrap` is passed with
+the parameter `--disable-tls`, the bootstrap configuration generated will make use of insecure
+communication between the containers.
+
+### Bootstrapping process
+
+This bootstrap phase is independent of `contour bootstrap` and uses `contour bootstrap-tls`.
+`contour bootstrap-tls` is run as an init container to bootstrap the TLS configuration.
+The bootstrap phase will create three pairs of the TLS certificates. The details about
+each pair of certificates are mentioned below:
+1. The first pair is used by Contour to serve HTTPS.
+2. The second pair is used by Envoy to authenticate with Contour.
+3. The last pair serves as the CA certificate which signs the above certificates.
+
+Each of these TLS certificates will be stored as Kubernetes secrets in namespace `heptio-contour`. The
+secrets will be named as `contour-tls`, `envoy-tls` and `contour-ca`, respectively. This ensures that
+even when the `contour` and `envoy` containers are not part of the same pod, the certificates are
+available for consumption by each of those.
+
+Note: In order for Contour to create the Kubernetes secrets in the `heptio-contour` namespace, the
+existing RBAC needs to be enhanced to grant permissions to `create` secrets in the namespace.
+
+### Setting up secure communication channel
+
+1. Contour container will pick the certificates from the Kubernetes secret `contour-tls` which is
+mounted at `/etc/tls/contour` and pick `contour-ca` mounted at `/etc/tls/ca` to start listening for
+HTTPS traffic.
+2. The bootstrap configuration yaml of Envoy will have [`tls_context`][1] added as part of `cluster` object.
+Envoy container will pick the certificates from the Kubernetes secret `envoy-tls` which is mounted at
+`/etc/tls/envoy` and pick `contour-ca` mounted at `/etc/tls/ca` to connect to Contour as configured by the
+bootstrap configuration file.
+
+### Rotation and invalidation of credentials
+
+This design assumes that to rotate or invalidate the credentials, Contour has to be redeployed which will
```

Comment:
What happens if I run 2 instances of Contour in the cluster? They will each overwrite the other's secrets. Maybe we need to have Contour check if the secrets exist, then they don't auto-generate. This would also allow users to pass their own certs if they wanted without using the auto-generated.

Reply:
Oh yeah, I need to consider this as well in the design. I will work on it.